### PR TITLE
File-local provisionals and explicit param declarations and overrides  

### PR DESCRIPTION
**Pull request by mandolaerik**
- **Use compat features to deprecate bad version statements**
- **Improve wording**
- **Use package_path from Simics base**
- **Refactor grammar for 'device'**
- **Add base syntax for provisional features**
- **Add syntax for walrus params**
- **Unify ASTs for param and typedparam**
- **Report errors for walrus violations**

### CODE REVIEW

**Comment by mandolaerik** [comment]
The commits from #277 were accidentally included (because I need some of them locally). Please review 277 before approving this one, to reduce noise.

**Comment by syssimics** [comment]
Verification [#13326740](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=13326740): fail

**Comment by syssimics** [comment]
Verification [#13337244](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=13337244): fail

**Comment by syssimics** [comment]
Verification [#13340724](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=13340724): fail

**Review by lwaern-intel — COMMENTED** [review]
This PR contains a lot of different changes, and I can't really tell which relate to what. How many of these do you need for the walrus params & introduction of provisionals? Are there any parts that can be split to a separate PR without one blocking the other? I'm especially curious to the many changes to the error tests you've made -- what are they about?

File: py/dml/dmlparse.py [hunk 1]
```diff
@@ -892,7 +900,8 @@ def object_statements_empty(t):
 @prod
 def object_statement(t):
     '''object_statement : object_statement_or_typedparam'''
-    if t[1].kind == 'typedparam':
+    if (t[1].kind == 'param' and t[1].args[1] is not None
+        and t[1].args[1].kind == 'paramtype'):
         report(ESYNTAX(site(t), ':',
```
Comment (lwaern-intel) [inline py/dml/dmlparse.py:905]: `s/':'/t[1].args[1]`

**Comment by syssimics** [comment]
Verification [#13346906](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=13346906): fail

File: py/dml/messages.py [hunk 2]
```diff
@@ -1138,6 +1138,25 @@ class EAUTOPARAM(DMLError):
     library, and they may not be overridden."""
     fmt = "bad declaration of automatic parameter '%s'"
 
+class ENWALRUS(DMLError):
+    """When the `walrus_params` provisional feature is enabled, parameter
+    declarations that do not override existing declarations must be
+    declared using the `:=` or `:default` syntax.
+    """
+    fmt = "non-overriding param should use :%s syntax (from walrus_params feature)"
```
Comment (lwaern-intel) [inline py/dml/messages.py:1146]: This error message is the **key feature** of walrus params, and the message should play to that.
```
fmt = "parameter not previously declared: '%s'. To declare and define a new parameter, use the ':%s' syntax (this requirement is due to the 'walrus_params' provisional feature being enabled)."
```

File: py/dml/messages.py [hunk 3]
```diff
@@ -1138,6 +1138,25 @@ class EAUTOPARAM(DMLError):
...
+
+class EWALRUS(DMLError):
+    """When the `walrus_params` provisional feature is enabled, parameters
+    declared using `:=` or `:default` syntax may not override other parameters.
+    """
+    fmt = "use of :%s to override existing param (from walrus_params feature)"
```
Comment (lwaern-intel) [inline py/dml/messages.py:1152]: `(the :%s syntax is provided by the walrus_params provisional feature)`, possibly with `... provisional feature, which has been enabled)`

File: py/dml/provisional.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+def feature(cls: type[ProvisionalFeature]):
+    assert issubclass(cls, ProvisionalFeature)
+    singleton = cls()
+    features[singleton.tag()] = singleton
+    return singleton
+
+
+@feature
+class walrus_params(ProvisionalFeature):
```
Comment (lwaern-intel) [inline py/dml/provisional.py:31]: I'm not fond of the name! Walrus params to me was an internal naming scheme.
Not that I'm not sure of a better name... maybe `explicit_param_overrides`? Or, if it's not too long, `explicit_param_decls_and_overrides`.

File: py/dml/messages.py [hunk 5]
```diff
@@ -1639,6 +1658,13 @@ class ELOGGROUPS(DMLError):
     fmt = ("Too many loggroup declarations. A maximum of 63 log groups (61 "
            + "excluding builtins) may be declared per device.")
 
+
+class ENOPROV(DMLError):
+    """
+    An invalid identifier was passed in the `provisional` statement.
+    """
+    fmt = "No such provisional feature %s. Valid values are: %s"
```
Comment (lwaern-intel) [inline py/dml/messages.py:1666]: I'd rather direct people to `--help-provisional`... once we've added that.

**Review by lwaern-intel — CHANGES_REQUESTED** [review]
Partial review before signing off, haven't looked into the implementation in depth. Will do a more in-depth review later.

File: py/dml/provisional.py (hunk 4 above)
Comment (lwaern-intel) [reply py/dml/provisional.py:31]: What am I thinking, the best would be `explicit_param_decls`.

File: py/dml/messages.py (hunk 3 above)
Comment (lwaern-intel) [reply py/dml/messages.py:1152]: By the way, I just realized that the following isn't illegal:
```
template x {
    param p = 4; // := with the new feature
}

template y is x {
    param p;
}

is y;
```
Perhaps it should be? I'm not sure. There's also the interplay with late shared, we don't want to make:
```
template x {
    param p = 4; // := with the new feature
}

template y is x {
    param p : int;
}

is y;
```
illegal.

Comment (lwaern-intel) [reply py/dml/messages.py:1152]: Errgh... late shared really does make this _weird_, doesn't it. Ideally we want `:=` to not be allowed to be used as an _override_, and yet...

```
template x {
    param p :default 4;
}

template y is x {
    param p : int;
    param p = 7;
}
```
Should be legal. If walruses truly are just syntactic sugar, why shouldn't
```
template y is x {
    param p : int = 7;
}
```
also be legal? And, by extension,

```
template y is x {
    param p := 7;
}
```
?

I think the answer should be that `:=` is only _mostly_ syntactic sugar, but does add an additional restriction that the multi-statement form does not: that the the param must not have been previously declared.

**Comment by mandolaerik** [comment]
What remains is documentation for provisionals, unclear if that blocks merge.

**Comment by syssimics** [comment]
Verification [#13399015](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=13399015): fail

**Comment by lwaern-intel** [comment]
> What remains is documentation for provisionals, unclear if that blocks merge.

And a passing verification!

When it comes to documentation, I'd like to see it both for provisionals as well as expl_param_decls (as per our R.C. discussions)

File: py/dml/messages.py (hunk 5 above)
Comment (mandolaerik) [reply py/dml/messages.py:1669]: I don't think we should add `--help-provisional`: provisionals are enabled only by syntax and thereby technically a part of the language, and describing the language is out of scope for help flags.

We may need to introduce flag-enabled provisionals later, but I hope not.

**Comment by syssimics** [comment]
Verification [#13401317](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=13401317): fail

File: py/dml/structure.py [hunk 6]
```diff
@@ -608,22 +636,47 @@ def is_default(decl):
...
+                    other for (other_rank, other) in decls + defs
+                    if other_rank in rank.inferior] + [
+                    other for (other_rank, other) in decls
+                    if other_rank is rank]
+                report(EOVERRIDE(type_info.site, type_info.site,#parent.site,
+                               'default' if is_default else '='))
+            elif (not is_override and declared_as_override
+                  and not any(r is rank for (r, _) in decls)):
+                report(ENOVERRIDE(
+                    p.site, 'default' if is_default else '='))
```
Comment (lwaern-intel) [inline py/dml/structure.py:664]: This entire loop strikes me as very messy. I think this could greatly benefit from the newly broken out `calc_minimal_ancestry`:
```
rank_to_decl_or_def = {r: spec for (r, spec) in itertools.chain(decls, defs) }
ancestry = calc_minimal_ancestry(Set(rank_to_decl_or_def))

for (rank, p) in defs:
    if p.site.provisional_enabled(provisional.explicit_param_decls):
        overrides = list(ancestry[rank])
        if not declared_as_override and overrides:
                report(EOVERRIDE(type_info.site, overrides[0], 'default' if is_default else '='))
        ....
```
... Actually, maybe the entirety of `merge_parameters` could be simplified by this.

File: py/dml/structure.py [hunk 7]
```diff
@@ -2711,11 +2766,12 @@ def mkexpr(self, indices):
...
 
 def mkparam(obj, autoparams, param):
-    _, site, name, (value, default, auto) = param
+    (_, site, name, type_info, is_default, value) = param
+    auto = type_info is not None and type_info.kind == 'auto'
 
     # auto-parameters are like normal "hard" assignments
     if name in autoparams:
-        if not auto and not default:
+        if not auto and not (dml.globals.dml_version == (1, 2) and is_default):
```
Comment (lwaern-intel) [inline py/dml/structure.py:2774]: If I'm interpreting this right, 1.2 abuses `default` declarations and possibly overrides them with autoparams? Wow.

File: test/1.4/provisional/T_bare.dml [hunk 8]
```diff
@@ -0,0 +1,17 @@
+/*
+  © 2024 Intel Corporation
+  SPDX-License-Identifier: MPL-2.0
+*/
+
+dml 1.4;
+
+/// COMPILE-ONLY
+
+provisional;
```
Comment (lwaern-intel) [inline test/1.4/provisional/T_bare.dml:10]: huh. I guess that makes sense. I just checked, Haskell permits a bare `{-# LANGUAGE #-}` too.

**Review by lwaern-intel — CHANGES_REQUESTED** [review]
Besides this, there are two things I want us to do, or at least come to an agreement about:
- Documentation
- How we would offer device-wide provisionals, as that would affect documentation

File: RELEASENOTES.docu [hunk 9]
```diff
@@ -1771,4 +1771,8 @@ extern typedef struct { } my_type_t;</pre> </add-note></build-id>
       m(conf_object_t)</tt> for function pointers is only permitted
       when this feature is enabled. The feature will be disabled in
       Simics API versions > 7.</add-note></build-id>
+  <build-id _6="next" _7="next"><add-note>Fixed a bug in
+      the <tt>dead_dml_methods</tt> library, which caused it to never
+      detect methods as dead if they occur directly after
+      a <tt>footer</tt> block.</add-note></build-id>
 </rn>
```
Comment (lwaern-intel) [inline RELEASENOTES.docu:1778]: We want a releasenote for the provisional mechanism, documenting its addition (because yes, it deserves it, in the same way `--compat` deserved it.) At the very least, we want it at the same time we offer documentation for provisionals, if we don't do that right away. Seperately we should have a releasenote for `explicit_param_decls`. (These both belong to 1.4's releasenotes.)

**Review by lwaern-intel — CHANGES_REQUESTED** [review]

**Comment by syssimics** [comment]
Verification [#13401431](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=13401431): pass

File: test/1.4/provisional/T_bare.dml (hunk 8 above)
Comment (mandolaerik) [reply test/1.4/provisional/T_bare.dml:10]: There is no intended use case, this is just a grammatical edge case; I added the test just to keep an unimportant corner stable.

File: py/dml/structure.py (hunk 7 above)
Comment (mandolaerik) [reply py/dml/structure.py:2774]: It's from `parameter indexvar default undefined;`.

BTW: This means that 1.2 incidentally allows user-provided definitions of autoparams, e.g.:
```
group g {
    parameter indexvar default 18;
    parameter bank default 3;
}
```
(where the indexvar decl will give `WEXPERIMENTAL: parameter with two default declarations` and have effect, whereas the bank decl will be silently swallowed and have no effect).

IIRC I once suggested to make it a proper auto but others (probably Jonatan) deemed that unimportant. And unimportant it is, but since you bothered to comment it I'll push another commit that cleans it up.

File: RELEASENOTES.docu (hunk 9 above)
Comment (mandolaerik) [reply RELEASENOTES.docu:1778]: A release note will come with its documentation.

File: py/dml/provisional.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+    In some rare cases, you may need to declare a parameter without
+    knowing if it's an override or a new parameter. In this case, one
+    can accompany a `param NAME = value;` or `param NAME default
+    value;` declaration with a `param NAME;` declaration in the same
+    scope/rank. This marks that the parameter assignment may be either
+    an override or a new parameter, and no error will be printed.
+
+    Enabling the `explicit_param_decls` feature in a file only affects
+    parameter declarations in that file.
```
[the remark this answers is not in the document]
Comment (mandolaerik) [reply py/dml/provisional.py:65]: My intent was to clarify the bounds of the checks; e.g., if a param is declared with `:=`, it's still OK to override it with `=` in a file that didn't enable the provisional.

File: py/dml/logging.py [hunk 11]
```diff
@@ -290,6 +292,8 @@ def dml_version(self):
         return self._dml_version
     def bitorder(self):
         return 'le'
+    def provisional_enabled(self, feature):
+        return False
```
[the remark this answers is not in the document]
Comment (mandolaerik) [reply py/dml/logging.py:296]: Why not? It's pretty harmless to assume that compiler-inserted statements use standard semantics. If this code path ever is taken and we have an ICE there, then I'm pretty sure we'll just insert `if feature in {...}: return False` to silence it, which won't contribute to our wisdom.

But I can surely ICE it if you insist.

File: py/dml/logging.py [hunk 12]
```diff
@@ -270,6 +270,8 @@ def lineno(self): pass
     @abc.abstractmethod
     def dml_version(self): pass
     @abc.abstractmethod
+    def provisional_enabled(self, feature): pass
```
[the remark this answers is not in the document]
Comment (mandolaerik) [reply py/dml/logging.py:273]: We can certainly let `provisional` keep them in a `frozenset` and return that. I don't think that would be significantly better or worse; I weakly prefer the current but can change if you insist.

**Comment by syssimics** [comment]
Verification [#13402763](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=13402763): pass

**Comment by syssimics** [comment]
Verification [#13403313](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=13403313): pass

File: py/dml/logging.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (lwaern-intel) [reply py/dml/logging.py:296]: We sometimes abuse `SimpleSite` for other purposes, don't we? That it's only used for compiler-inserted statements, I think, is at least not something we've firmly established.

[the remark this answers is not in the document]
Comment (mandolaerik) [reply py/dml/logging.py:296]: It should be established that statements that come from DML files don't use SimpleSite. And file-local provisionals are only applicable to statements from DML files. E.g., SimpleSite is used for user-defined expressions like `-D` (`SimpleSite('<command-line>')`), where file-local provisionals should not apply.

File: py/dml/logging.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (lwaern-intel) [reply py/dml/logging.py:273]: Naah. Do as you prefer.

File: py/dml/logging.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (lwaern-intel) [reply py/dml/logging.py:296]: Looking through the usages of `SimpleSite` in DMLC, I find that questionable. I'd prefer you ICE it unless it turns it gets triggered in a way that can only be fixed by suprressing it at a centralized site that renders the ICE effectively worthless.

**Comment by syssimics** [comment]
Verification [#13411421](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=13411421): ✅ pass

File: py/dml/structure.py (hunk 6 above)
Comment (lwaern-intel) [reply py/dml/structure.py:668]: @mandolaerik I've seen your new code -- how come you are still clumsily determining a parent via a list comprehension? I don't get it. Why would you not use the same or a similar pattern as I exemplify in this comment?

Comment (lwaern-intel) [reply py/dml/structure.py:668]: oh, maybe i'm confusing. In the above comment code: `s/overrides/overridden`. I meant in the sense "that which p overrides", but that is not a good choice of words

Comment (mandolaerik) [reply py/dml/structure.py:668]: The strongest reason is that `list(ancestry[rank])[0]` depends on set iteration order. Also, your suggestion reduces to something like:
```
parent = dict(itertools.chain(decls.items(), defs.items()))[next(iter(minimal_ancestry[rank]))]
```
which I find harder to read.

Comment (lwaern-intel) [reply py/dml/structure.py:668]: With the code as-is you might as well not use `calc_minimal_ancestry` at all. I don't buy the "if don't like it cause if I minimize the code it becomes ugly" argument; then don't minimize the code. This does not affect generated C, so the undefined set iteration order argument is weak; but does carry the most weight.
What about reworking `Rank` to operate in terms of `Set`? Would the induced slowdown be too much? If so, if you insist on not using the values of the map `calc_minimal_ancestry` gives you, then don't use the function at all. I still feel the loop, both before and as-is (because as-is is just the exact same pattern & algorithm you employed before, just with `calc_minimal_ancestry` shoe-horned in) is _hideously_ ugly, and I would **really** want it improved, but if you can't find a way then so be it.

Comment (mandolaerik) [reply py/dml/structure.py:668]: Sorry, I don't see how your suggestion would improve the situation.
```
is_override = not all_ranks.isdisjoint(rank.inferior
```
vs
```
is_override = bool(calc_minimal_ancestry(all_ranks)[rank])
```
and
```
                minimal_ancestry = traits.calc_minimal_ancestry(all_ranks)
                [parent, *_] = (parent for (parent_rank, parent) in params
                                if parent_rank in minimal_ancestry[rank])
```
vs
```
                rank_to_decl_or_def = dict(itertools.chain(defs, decls))
                all_overrides = traits.calc_minimal_ancestry(all_ranks)
                [one_override, *] = all_overrides
                parent = rank_to_decl_or_def[one_override]
```
I don't understand why the latter would be better code.

Comment (lwaern-intel) [reply py/dml/structure.py:668]: Because
```
rank_to_decl_or_def = dict(itertools.chain(defs, decls))
all_overrides = traits.calc_minimal_ancestry(all_ranks)
```
Doesn't have to be part of the loop? You don't have to, you know, _recalculate_ `calc_minimal_ancestry` for every param def/decl. And I find avoiding the need to clumsily spell out `(parent for (parent_rank, parent) in params if parent_rank in minimal_ancestry[rank])` a benefit.

Comment (lwaern-intel) [reply py/dml/structure.py:668]:
```
all_overrides = traits.calc_minimal_ancestry(all_ranks)
[one_override, *] = all_overrides
```
isn't even legal. Are you simply misunderstanding what `calc_minimal_ancestry` returns?

Comment (lwaern-intel) [reply py/dml/structure.py:668]: In particular, your example of `is_override = bool(calc_minimal_ancestry(all_ranks)[rank])` is disingenuous because you can just have the result of `calc_minimal_ancestry(all_ranks)` be calculated before the loop and shared between each iteration of it, reducing the check to `is_override = bool(minimal_ancestry[rank])`.

File: RELEASENOTES-1.4.docu [hunk 13]
```diff
@@ -558,4 +558,17 @@
       in multiple inheritance: it allows any method that overrides multiple
       implementations from unrelated templates to individually reference and
       call the implementations it overrides.</add-note></build-id>
+  <build-id _6="next" _7="next"><add-note>Added a system
+      for <em>provisional language features</em>: language features
+      that are disabled by default but can be explicitly enabled with
+      a special syntax. The first provisional feature
+      is <tt>explicit_param_decls</tt>, which can be enabled per file
```
Comment (lwaern-intel) [inline RELEASENOTES-1.4.docu:565]: You should split these up to separate releasenotes. One for the provisional system, one for the `explicit_param_decls`. By the way, don't provisionals have a JIRA of their own, or did we never create one?

**Review by lwaern-intel — CHANGES_REQUESTED** [review]

File: py/dml/dmlparse.py (hunk 1 above)
Comment (mandolaerik) [reply py/dml/dmlparse.py]: Chose `None` instead

File: py/dml/messages.py (hunk 2 above)
Comment (mandolaerik) [reply py/dml/messages.py]: Improved. I chose to include the provisional declaration on a separate context line.

File: py/dml/messages.py (hunk 3 above)
Comment (mandolaerik) [reply py/dml/messages.py]: Regarding your original comment: I chose instead to not mention the provisional feature. If the walrus provisional is enabled in a file and you write `:=` in that file, then you are aware of walruses, so that information is redundant. It would make more sense to add a walrus hint to the `ESYNTAX` for `:=`, we can do that when the provisional goes stable.

File: py/dml/provisional.py (hunk 4 above)
Comment (mandolaerik) [reply py/dml/provisional.py]: changed!

File: py/dml/logging.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (mandolaerik) [reply py/dml/logging.py:296]: discussed in RC; this is similar to `bitorder` and `dml_version`, where a ICE in SimpleSite would not have gained us.

File: py/dml/structure.py [hunk 14]
```diff
@@ -603,22 +631,47 @@ def is_default(decl):
...
+            is_override = (not decl_or_def_ranks.isdisjoint(rank.inferior)
+                           or rank in decl_ranks)
+            if is_walrus and is_override:
+                [parent, *_] = [
+                    other for (other_rank, other) in decls + defs
+                    if other_rank in rank.inferior] + [
+                    other for (other_rank, other) in decls
+                    if other_rank is rank]
+                report(EWALRUS(type_info.site, type_info.site,#parent.site,
+                               'default' if is_default else '='))
```
[the remark this answers is not in the document]
Comment (mandolaerik) [reply py/dml/structure.py]: fixed

File: py/dml/provisional.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (mandolaerik) [reply py/dml/provisional.py:68]: Wrote something to clarify this.

File: py/dml/structure.py (hunk 6 above)
Comment (mandolaerik) [reply py/dml/structure.py:668]: Things are somewhat complex, there are performance concerns. I solved it by memoizing calc_minimal_ancestry.

File: RELEASENOTES.docu (hunk 9 above)
Comment (mandolaerik) [reply RELEASENOTES.docu:1778]: A release note has been written.

File: RELEASENOTES-1.4.docu (hunk 13 above)
Comment (mandolaerik) [reply RELEASENOTES-1.4.docu:565]: I didn't find a jira from a quick search.

I have split the releasenote.

**Comment by syssimics** [comment]
Verification [#13429364](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=13429364): ✅ pass

File: py/dml/structure.py [hunk 15]
```diff
@@ -542,88 +557,130 @@ def add_templates(obj_specs, each_stmts):
...
+            if not declared_as_override and is_override:
+                [parent, *_] = (parent for (parent_rank, parent) in params
+                                if parent_rank in minimal_ancestry[rank])
+                report(EOVERRIDE(type_info.site, parent.site,
+                                 'default' if is_default else '='))
+            if not declared_as_override and rank in decls:
+                report(EOVERRIDE(type_info.site, decls[rank].site,
+                                 'default' if is_default else '='))
+            elif (not is_override and declared_as_override
+                  and rank not in decls):
```
Comment (lwaern-intel) [inline py/dml/structure.py:666]: I slightly prefer:
```
parents = minimal_ancestry[rank]
if not declared_as_override and parents:
                [parent, *_] = (parent for (parent_rank, parent) in params
                                if parent_rank in parents)
...
```
But I don't insist.

File: RELEASENOTES-1.4.docu (hunk 13 above)
Comment (lwaern-intel) [reply RELEASENOTES-1.4.docu:568]: If you have, you haven't pushed that up!

**Review by lwaern-intel — CHANGES_REQUESTED** [review]
Approved modulo a small thing that doesn't block merge, and the fact that you seem to have a commit unpushed, which does block merge.

File: py/dml/messages.py (hunk 5 above)
Comment (lwaern-intel) [reply py/dml/messages.py:1678]: Discussed outside R.C.: he does not, and there are some good arguments for not liking it. How we'll approach device-wide provisionals will have to be resolved later.

**Comment by mandolaerik** [comment]
> If you have, you haven't pushed that up!

Oops, for unknown reason the split was moved to my editor's undo history. Fixed.

File: RELEASENOTES-1.4.docu [hunk 16]
```diff
@@ -559,6 +559,20 @@
...
       call the implementations it overrides.</add-note></build-id>
   <build-id _6="6309" _7="7021"><add-note> Template <tt>function_io_memory</tt>
-      now searches for the bank in the same `subdevice`/`device` scope as the
-      `implement`. </add-note></build-id>
+      now searches for the bank in the same <tt>subdevice</tt>/<tt>device</tt> scope as the
+      <tt>implement</tt>. </add-note></build-id>
+  <build-id _6="next" _7="next"><add-note>Added a system
+      for <em>provisional language features</em>: language features
+      that are disabled by default but can be explicitly enabled with
+      a special syntax. </add-note></build-id>
```
Comment (lwaern-intel) [inline RELEASENOTES-1.4.docu:567]:
```
can be explicitly enabled per file via the new <tt>provisional</tt> statement,
which may only occur right after the DML version specification.
```

**Review by lwaern-intel — CHANGES_REQUESTED** [review]

File: RELEASENOTES-1.4.docu (hunk 16 above)
Comment (lwaern-intel) [reply RELEASENOTES-1.4.docu:567]: At the very least, that it's per-file is important to note.

**Comment by syssimics** [comment]
Verification [#13431918](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=13431918): ✅ pass

**Comment by syssimics** [comment]
Verification [#13432062](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=13432062): ✅ pass

File: RELEASENOTES-1.4.docu (hunk 16 above)
Comment (mandolaerik) [reply RELEASENOTES-1.4.docu]: Fixed.

**Review by lwaern-intel — APPROVED** [review]
Approved! You should however squash together some of the 15 commits.

**Comment by syssimics** [comment]
Verification [#13432352](https://teamcity01-ir.devtools.intel.com/viewLog.html?buildId=13432352): ✅ pass

**Comment by mandolaerik** [comment]
I don't think squashing is worth the effort, only ~2 of the commits are squashable.
